README.md
# harbor_pay

A Flutter package to accept mobile money payments.

## Getting Started

First, add the `harbor_pay` package to your pubspec dependencies.

### Import HarborPay

```dart
import 'package:harbor_pay/harbor_pay.dart';
```

### Create a HarborPay object

```dart

HPay hp = new HPay(
    clientId: '123456',
    buttonColors: Colors.black,
    bottomSheetBackgroundColor: Colors.white
);
```

### Process Payment

```dart

var paymentResponse = await
hp.processPayment(context: context, amount: 14.00
,
customerNumber: '
2330000000
'
);
print
(
paymentResponse.toString()
);

// Optional but useful parameter: customerName
// Optional but useful parameter: Map<String, dynamic> extra
```

`paymentResponse` from the calls, will contain the following in JSON response:

```json
{
  "success": true,
  "message": ""
}
```

A `success` value of `true` means the payment has actually been processed successfully and no
further action has to be taken.

CHANGELOG.md
## [1.0.1] - 2023-10-21
- Revamped to use default base URL

## [1.0.0] - 2022-08-11
- Revamped to only accept payments using mobile money

- ## [0.0.9] - 2021-04-14
- Relaxed customerName requirement for processing payments

## [0.0.8] - 2021-04-13
- Migrated to null safety.
- Added source to initializer.

## [0.0.7] - 2021-03-22
- Corrected layout issue with PayPal.

## [0.0.6] - 2021-03-19
- Added PayPal to the payment methods.

## [0.0.5] - 2021-02-15
- Fixed mobile provider selection issues.

## [0.0.4] - 2021-02-12
- Added option to withdraw money from your HarborPay account to a mobile wallet.

## [0.0.3] - 2021-02-09
- Added option to send money to multiple recipients from your HarborPay account.

## [0.0.2] - 2021-02-05
- Added deposit functionality to deposit money into your HarborPay account.
- Added sendMoney functionality to send money from your HarborPay account.

## [0.0.1] - 2021-02-03
The initial release supports mobile money payments only.

lib/hp_io.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

const base = "excelliumgh.com";
const appLightGray = Color(0xFFE7ECF2);
const appGray = Color(0xFFadb4b9);
const appDarkText = Color(0xFF2F2F2F);

const BASE_SHEET_HEIGHT = 160.00;
const MOMO_SHEET_HEIGHT = 180.00;
const MOMO_PAY_METHOD = 'momo';
const CARD_PAY_METHOD = 'card';
const MOMO_MTN = 'MTN';
const MOMO_VODAFONE = 'VOD';
const MOMO_AIRTELTIGO = 'AIR';

const META_PURPOSE_DEPOSIT = 'deposit';

String payMethod = CARD_PAY_METHOD;
String selectedMomo = MOMO_MTN;

doPost(String urlAfterBase, Map bod, {Map<String, String>? headers}) async {
  var body = stripNulls(bod);
  var uri = Uri.https(base, '/auto/api/$urlAfterBase');
  var js = await http.post(uri, body: jsonEncode(body), headers: headers);
  var decoded = jsonDecode(js.body);
  return decoded;
}

doGet(String urlAfterBase, {Map<String, dynamic>? queries, Map<String, String>? headers}) async {
  var uri = Uri.https(base, '/auto/api/$urlAfterBase', queries);
  var js = await http.get(uri, headers: headers);
  var decoded;
  try {
    decoded = jsonDecode(js.body);
  } catch (e) {
    print(e);
  }
  return decoded;
}

Map stripNulls(Map m) {
  Map finMap = {};
  for (var i in m.keys) {
    if (m[i] is String) {
      if (m[i] != '') {
        finMap[i] = m[i];
      }
    } else if (m[i] == null) {
    } else if (m[i] is double) {
      finMap[i] = m[i].toString();
    } else {
      finMap[i] = jsonEncode(m[i]);
    }
  }
  return finMap;
}

Widget appText(String word, double z, FontWeight w,
    {Color col = appDarkText, TextAlign align = TextAlign.left, int maxLines = 5, int shadow = 0}) {
  return Text(
    word,
    softWrap: true,
    maxLines: maxLines,
    overflow: TextOverflow.ellipsis,
    textAlign: align,
    style: TextStyle(
      color: col,
      fontFamily: 'Open Sans',
      fontSize: z,
      fontWeight: w,
      shadows: shadow > 0 ? elevation(Colors.black38, shadow) : [],
    ),
  );
}

InputDecoration textDecor(
    {String? hint,
    Icon? prefixIcon,
    Widget? suffix,
    String prefix = '',
    bool enabled = true,
    double hintSize = 16,
    bool showBorder = true}) {
  return new InputDecoration(
    prefixIcon: prefixIcon,
    prefixText: prefix,
    suffix: suffix,
    labelText: hint,
    hintStyle: TextStyle(fontSize: hintSize, color: appGray),
    enabledBorder: OutlineInputBorder(
      borderSide: BorderSide(color: appDarkText.withOpacity(0.1), width: 1),
      borderRadius: BorderRadius.circular(4),
    ),
    focusedBorder: OutlineInputBorder(
      borderSide: BorderSide(color: appDarkText.withOpacity(0.1), width: 1),
      borderRadius: BorderRadius.circular(4),
    ),
    labelStyle: TextStyle(fontSize: hintSize, color: appGray),
    filled: true,
    fillColor: Colors.white,
    contentPadding: EdgeInsets.fromLTRB(20.0, 12.0, 20.0, 12.0),
  );
}

ButtonStyle bStyle(Color color, {double radius: 0}) {
  return ButtonStyle(
      backgroundColor: MaterialStateProperty.all(color),
      shape: MaterialStateProperty.all(RoundedRectangleBorder(borderRadius: BorderRadius.circular(radius))));
}

List<BoxShadow> elevation(Color col, int elevation) {
  return [
    BoxShadow(color: col.withOpacity(0.6), offset: Offset(0.0, 4.0), blurRadius: 3.0 * elevation, spreadRadius: -1.0 * elevation),
    BoxShadow(color: col.withOpacity(0.44), offset: Offset(0.0, 1.0), blurRadius: 2.2 * elevation, spreadRadius: 1.5),
    BoxShadow(color: col.withOpacity(0.12), offset: Offset(0.0, 1.0), blurRadius: 4.6 * elevation, spreadRadius: 0.0),
  ];
}

Widget progress({double size = 30}) {
  return SizedBox(
      width: size,
      height: size,
      child: CircularProgressIndicator(backgroundColor: appLightGray, valueColor: const AlwaysStoppedAnimation<Color>(Color(0xFF9c9c9c))));
}

double appWidth(con) {
  return MediaQuery.of(con).size.width;
}

double appHeight(con) {
  return MediaQuery.of(con).size.height;
}

EdgeInsets verticalPadding(double size) {
  return EdgeInsets.symmetric(vertical: size);
}


lib/bsheet.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

const Duration _kBottomSheetDuration = Duration(milliseconds: 200);
const double _kMinFlingVelocity = 700.0;
const double _kCloseProgressThreshold = 0.5;

/// A material design bottom sheet.
///
/// There are two kinds of bottom sheets in material design:
///
///  * _Persistent_. A persistent bottom sheet shows information that
///    supplements the primary content of the app. A persistent bottom sheet
///    remains visible even when the user interacts with other parts of the app.
///    Persistent bottom sheets can be created and displayed with the
///    [ScaffoldState.showBottomSheet] function or by specifying the
///    [Scaffold.bottomSheet] constructor parameter.
///
///  * _Modal_. A modal bottom sheet is an alternative to a menu or a dialog and
///    prevents the user from interacting with the rest of the app. Modal bottom
///    sheets can be created and displayed with the [showModalBottomSheet]
///    function.
///
/// The [BottomSheet] widget itself is rarely used directly. Instead, prefer to
/// create a persistent bottom sheet with [ScaffoldState.showBottomSheet] or
/// [Scaffold.bottomSheet], and a modal bottom sheet with [showModalBottomSheet].
///
/// See also:
///
///  * [showBottomSheet] and [ScaffoldState.showBottomSheet], for showing
///    non-modal "persistent" bottom sheets.
///  * [showModalBottomSheet], which can be used to display a modal bottom
///    sheet.
///  * <https://material.io/design/components/sheets-bottom.html>
class BottomSheet extends StatefulWidget {
  /// Creates a bottom sheet.
  ///
  /// Typically, bottom sheets are created implicitly by
  /// [ScaffoldState.showBottomSheet], for persistent bottom sheets, or by
  /// [showModalBottomSheet], for modal bottom sheets.
  const BottomSheet({
    Key key,
    this.animationController,
    this.enableDrag = true,
    this.elevation = 0.0,
    @required this.onClosing,
    @required this.builder,
  })  : assert(enableDrag != null),
        assert(onClosing != null),
        assert(builder != null),
        assert(elevation != null && elevation >= 0.0),
        super(key: key);

  /// The animation that controls the bottom sheet's position.
  ///
  /// The BottomSheet widget will manipulate the position of this animation, it
  /// is not just a passive observer.
  final AnimationController animationController;

  /// Called when the bottom sheet begins to close.
  ///
  /// A bottom sheet might be prevented from closing (e.g., by user
  /// interaction) even after this callback is called. For this reason, this
  /// callback might be call multiple times for a given bottom sheet.
  final VoidCallback onClosing;

  /// A builder for the contents of the sheet.
  ///
  /// The bottom sheet will wrap the widget produced by this builder in a
  /// [Material] widget.
  final WidgetBuilder builder;

  /// If true, the bottom sheet can dragged up and down and dismissed by swiping
  /// downwards.
  ///
  /// Default is true.
  final bool enableDrag;

  /// The z-coordinate at which to place this material relative to its parent.
  ///
  /// This controls the size of the shadow below the material.
  ///
  /// Defaults to 0. The value is non-negative.
  final double elevation;

  @override
  _BottomSheetState createState() => _BottomSheetState();

  /// Creates an animation controller suitable for controlling a [BottomSheet].
  static AnimationController createAnimationController(TickerProvider vsync) {
    return AnimationController(
      duration: _kBottomSheetDuration,
      debugLabel: 'BottomSheet',
      vsync: vsync,
    );
  }
}

class _BottomSheetState extends State<BottomSheet> {
  final GlobalKey _childKey = GlobalKey(debugLabel: 'BottomSheet child');

  double get _childHeight {
    final RenderBox renderBox = _childKey.currentContext.findRenderObject();
    return renderBox.size.height;
  }

  bool get _dismissUnderway => widget.animationController.status == AnimationStatus.reverse;

  void _handleDragUpdate(DragUpdateDetails details) {
    if (_dismissUnderway) return;
    widget.animationController.value -= details.primaryDelta / (_childHeight ?? details.primaryDelta);
  }

  void _handleDragEnd(DragEndDetails details) {
    if (_dismissUnderway) return;
    if (details.velocity.pixelsPerSecond.dy > _kMinFlingVelocity) {
      final double flingVelocity = -details.velocity.pixelsPerSecond.dy / _childHeight;
      if (widget.animationController.value > 0.0) widget.animationController.fling(velocity: flingVelocity);
      if (flingVelocity < 0.0) widget.onClosing();
    } else if (widget.animationController.value < _kCloseProgressThreshold) {
      if (widget.animationController.value > 0.0) widget.animationController.fling(velocity: -1.0);
      widget.onClosing();
    } else {
      widget.animationController.forward();
    }
  }

  @override
  Widget build(BuildContext context) {
    final Widget bottomSheet = Material(
      key: _childKey,
      elevation: widget.elevation,
      child: widget.builder(context),
    );
    return !widget.enableDrag
        ? bottomSheet
        : GestureDetector(
            onVerticalDragUpdate: _handleDragUpdate,
            onVerticalDragEnd: _handleDragEnd,
            child: ClipRRect(borderRadius: BorderRadius.circular(10), child: bottomSheet),
            excludeFromSemantics: true,
          );
  }
}

// PERSISTENT BOTTOM SHEETS

// See scaffold.dart

// MODAL BOTTOM SHEETS

class _ModalBottomSheetLayout extends SingleChildLayoutDelegate {
  _ModalBottomSheetLayout(this.progress);

  final double progress;

  @override
  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
    return BoxConstraints(
      minWidth: constraints.maxWidth,
      maxWidth: constraints.maxWidth,
      minHeight: 0.0,
      maxHeight: constraints.maxHeight,
    );
  }

  @override
  Offset getPositionForChild(Size size, Size childSize) {
    return Offset(0.0, size.height - childSize.height * progress);
  }

  @override
  bool shouldRelayout(_ModalBottomSheetLayout oldDelegate) {
    return progress != oldDelegate.progress;
  }
}

class _ModalBottomSheet<T> extends StatefulWidget {
  const _ModalBottomSheet({Key key, this.route}) : super(key: key);

  final _ModalBottomSheetRoute<T> route;

  @override
  _ModalBottomSheetState<T> createState() => _ModalBottomSheetState<T>();
}

class _ModalBottomSheetState<T> extends State<_ModalBottomSheet<T>> {
  @override
  Widget build(BuildContext context) {
    final MediaQueryData mediaQuery = MediaQuery.of(context);
    final MaterialLocalizations localizations = MaterialLocalizations.of(context);
    String routeLabel;
    switch (defaultTargetPlatform) {
      case TargetPlatform.windows:
      case TargetPlatform.macOS:
      case TargetPlatform.linux:
      case TargetPlatform.iOS:
        routeLabel = '';
        break;
      case TargetPlatform.android:
      case TargetPlatform.fuchsia:
        routeLabel = localizations.dialogLabel;
        break;
    }

    return GestureDetector(
      excludeFromSemantics: true,
      onTap: () => Navigator.pop(context),
      child: AnimatedBuilder(
        animation: widget.route.animation,
        builder: (BuildContext context, Widget child) {
          // Disable the initial animation when accessible navigation is on so
          // that the semantics are added to the tree at the correct time.
          final double animationValue = mediaQuery.accessibleNavigation ? 1.0 : widget.route.animation.value;
          return Semantics(
            scopesRoute: true,
            namesRoute: true,
            label: routeLabel,
            explicitChildNodes: true,
            child: ClipRect(
              child: CustomSingleChildLayout(
                delegate: _ModalBottomSheetLayout(animationValue),
                child: BottomSheet(
                  animationController: widget.route._animationController,
                  onClosing: () => Navigator.pop(context),
                  builder: widget.route.builder,
                ),
              ),
            ),
          );
        },
      ),
    );
  }
}

class _ModalBottomSheetRoute<T> extends PopupRoute<T> {
  _ModalBottomSheetRoute({
    this.builder,
    this.theme,
    this.barrierLabel,
    RouteSettings settings,
  }) : super(settings: settings);

  final WidgetBuilder builder;
  final ThemeData theme;

  @override
  Duration get transitionDuration => _kBottomSheetDuration;

  @override
  bool get barrierDismissible => true;

  @override
  final String barrierLabel;

  @override
  Color get barrierColor => Colors.black54;

  AnimationController _animationController;

  @override
  AnimationController createAnimationController() {
    assert(_animationController == null);
    _animationController = BottomSheet.createAnimationController(navigator.overlay);
    return _animationController;
  }

  @override
  Widget buildPage(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) {
    // By definition, the bottom sheet is aligned to the bottom of the page
    // and isn't exposed to the top padding of the MediaQuery.
    Widget bottomSheet = MediaQuery.removePadding(
      context: context,
      removeTop: true,
      child: _ModalBottomSheet<T>(route: this),
    );
    if (theme != null) bottomSheet = Theme(data: theme, child: bottomSheet);
    return bottomSheet;
  }
}

/// Shows a modal material design bottom sheet.
///
/// A modal bottom sheet is an alternative to a menu or a dialog and prevents
/// the user from interacting with the rest of the app.
///
/// A closely related widget is a persistent bottom sheet, which shows
/// information that supplements the primary content of the app without
/// preventing the use from interacting with the app. Persistent bottom sheets
/// can be created and displayed with the [showBottomSheet] function or the
/// [ScaffoldState.showBottomSheet] method.
///
/// The `context` argument is used to look up the [Navigator] and [Theme] for
/// the bottom sheet. It is only used when the method is called. Its
/// corresponding widget can be safely removed from the tree before the bottom
/// sheet is closed.
///
/// Returns a `Future` that resolves to the value (if any) that was passed to
/// [Navigator.pop] when the modal bottom sheet was closed.
///
/// See also:
///
///  * [BottomSheet], which is the widget normally returned by the function
///    passed as the `builder` argument to [showModalBottomSheet].
///  * [showBottomSheet] and [ScaffoldState.showBottomSheet], for showing
///    non-modal bottom sheets.
///  * <https://material.io/design/components/sheets-bottom.html#modal-bottom-sheet>
Future<T> showModalBottomSheetCustom<T>({
  @required BuildContext context,
  @required WidgetBuilder builder,
}) {
  assert(context != null);
  assert(builder != null);
  assert(debugCheckHasMaterialLocalizations(context));
  return Navigator.push(
      context,
      _ModalBottomSheetRoute<T>(
        builder: builder,
        theme: Theme.of(context),
        barrierLabel: MaterialLocalizations.of(context).modalBarrierDismissLabel,
      ));
}

/// Shows a persistent material design bottom sheet in the nearest [Scaffold].
///
/// Returns a controller that can be used to close and otherwise manipulate the
/// bottom sheet.
///
/// To rebuild the bottom sheet (e.g. if it is stateful), call
/// [PersistentBottomSheetController.setState] on the controller returned by
/// this method.
///
/// The new bottom sheet becomes a [LocalHistoryEntry] for the enclosing
/// [ModalRoute] and a back button is added to the appbar of the [Scaffold]
/// that closes the bottom sheet.
///
/// To create a persistent bottom sheet that is not a [LocalHistoryEntry] and
/// does not add a back button to the enclosing Scaffold's appbar, use the
/// [Scaffold.bottomSheet] constructor parameter.
///
/// A persistent bottom sheet shows information that supplements the primary
/// content of the app. A persistent bottom sheet remains visible even when
/// the user interacts with other parts of the app.
///
/// A closely related widget is a modal bottom sheet, which is an alternative
/// to a menu or a dialog and prevents the user from interacting with the rest
/// of the app. Modal bottom sheets can be created and displayed with the
/// [showModalBottomSheet] function.
///
/// The `context` argument is used to look up the [Scaffold] for the bottom
/// sheet. It is only used when the method is called. Its corresponding widget
/// can be safely removed from the tree before the bottom sheet is closed.
///
/// See also:
///
///  * [BottomSheet], which is the widget typically returned by the `builder`.
///  * [showModalBottomSheet], which can be used to display a modal bottom
///    sheet.
///  * [Scaffold.of], for information about how to obtain the [BuildContext].
///  * <https://material.io/design/components/sheets-bottom.html#standard-bottom-sheet>
PersistentBottomSheetController<T> showBottomSheet<T>({
  @required BuildContext context,
  @required WidgetBuilder builder,
}) {
  assert(context != null);
  assert(builder != null);
  return Scaffold.of(context).showBottomSheet<T>(builder);
}


lib/src/hp.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:harbor_pay/hp_io.dart';

class HPay {
  HPay(
      {required this.clientId,
      required this.callback,
      this.currency = 'GHS',
      this.purpose = '',
      this.reference = '',
      this.transactionType = 'momo',
      this.customerNumber = '',
      this.buttonColors = const Color(0xFF24b4c4),
      this.bottomSheetBackgroundColor = const Color(0xff787878)});

  String clientId;
  String currency;
  double amount = 0;
  String reference; //  used as transaction id. If not passed, a random string will be generated
  String purpose; //  useful when topping up a HarborPay wallet
  String callback; // where to receive status updates
  String transactionType; //  momo or card
  String transactionPlatform = 'MTN'; //AIR, MTN or VOD;
  String customerNumber;
  Color buttonColors;
  Color bottomSheetBackgroundColor;

  genId({int count = 8}) {
    var r = Random();
    const _chars = 'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890';
    return List.generate(count, (index) => _chars[r.nextInt(_chars.length)]).join();
  }

  initialize(HPay options) {
    this.clientId = options.clientId;
    this.amount = options.amount;
    this.currency = options.currency;
    this.reference = options.reference.isEmpty ? this.genId() : options.reference;
    this.purpose = options.purpose.isEmpty ? '{}' : options.purpose;
  }

  appendToInitializer(options) {
    if (options.transactionType != null) this.transactionType = options.transactionType;
    if (options.transactionPlatform != null) this.transactionPlatform = options.transactionPlatform;
    if (options.customerNumber != null) this.customerNumber = options.customerNumber;
    if (options.currency) this.currency = options.currency;
    if (options.amount) this.amount = options.amount;
    if (options.purpose) this.purpose = options.purpose;
  }

  prepareTransaction(context, {Map extra: const {}}) async {
    if (!['momo', 'card'].contains(this.transactionType)) {
      return {
        'success': false,
        'message': 'parameter transactionType should be one of [\'momo\', \'card\']',
      };
    }
    if (this.amount <= 0) {
      return {
        'success': false,
        'message': 'amount should be more than zero',
      };
    }

    var finalLoad = {
      "client": this.clientId,
      "reference": this.reference != '' ? this.reference : genId(count: 10),
      "transaction_type": this.transactionType,
      "customer_number": this.customerNumber,
      "currency": this.currency,
      "amount": this.amount,
      "purpose": this.purpose,
      "callback": this.callback
    };

    if (selectedMomo != '') {
      this.transactionPlatform = selectedMomo;
      finalLoad['network_type'] = this.transactionPlatform;
    }

    if (this.purpose != '') {
      if (['deposit', 'top_up'].indexOf(this.purpose) > 0) {
        finalLoad['meta'] = {'purpose': this.purpose, 'client': this.clientId};
      }
    }

    if (extra.length > 0) finalLoad['meta'] = extra;

//    print(finalLoad.toString());

    var x = await doPost('forward', finalLoad, headers: headers());

    if (x['resp_code'] == '015') {
      var responseMap = await queryTransStatus(finalLoad['reference']);
      Navigator.pop(context, responseMap);
    } else {
      Navigator.pop(context, x);
    }
  }

  Future<Map<String, dynamic>> queryTransStatus(transId) async {
    for (int i = 0; i < 5; i++) {
      await Future.delayed(Duration(seconds: 6));
      var dec = await doPost('verify', {'client': this.clientId, 'reference': transId}, headers: headers());
      if (dec['transactionStatus'] == 'Completed') {
        return {'success': true, 'message': 'Payment Received. Thank you'};
      } else if (dec['transactionStatus'] == 'Failed') {
        return {'success': false, 'message': 'Payment Failed. User did not confirm'};
      }
    }
    return {'success': false, 'message': 'Payment Failed. Timed out'};
  }

  Future processPayment(
      {required BuildContext context,
      required double amount,
      String? customerName,
      required String customerNumber,
      String? purpose,
      Map extra = const {}}) async {
    if (amount != 0) this.amount = amount;
    if (customerNumber != '') this.customerNumber = customerNumber;
    if (purpose != null) this.purpose = purpose;

    int currentStep = 0;
    double height = 340;

    return await showModalBottomSheet(
        isScrollControlled: true,
        context: context,
        builder: (_) {
          return StatefulBuilder(
              builder: (BuildContext context, StateSetter setState) => Container(
                  height: height,
                  padding: EdgeInsets.all(10.0),
                  decoration: BoxDecoration(
                    color: this.bottomSheetBackgroundColor,
                  ),
                  child: Stepper(
                    currentStep: currentStep,
                    type: StepperType.horizontal,
                    onStepContinue: () {
                      setState(() => currentStep = currentStep + 1);
                    },
                    onStepCancel: () {
                      setState(() => currentStep = currentStep - 1);
                    },
                    steps: [
                      Step(
                        isActive: currentStep == 0,
                        title: Text('Provider'),
                        content: Container(
                            height: 300,
                            child: ListView(
                              children: [
                                Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    appText('Payment for:', 12, FontWeight.normal),
                                  ],
                                ),
                                Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    Padding(
                                      padding: const EdgeInsets.all(8.0),
                                      child: appText('${this.currency}${this.amount.toString()}', 30, FontWeight.bold),
                                    ),
                                  ],
                                ),
                                Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    appText('Select mobile money platform:', 12, FontWeight.normal),
                                  ],
                                ),
                                Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    Container(
                                      margin: verticalPadding(10),
                                      width: appWidth(context) * 0.6,
                                      decoration: BoxDecoration(
                                        color: appLightGray.withOpacity(0.5),
                                        borderRadius: BorderRadius.circular(40),
                                      ),
                                      child: Container(
                                        padding: EdgeInsets.symmetric(horizontal: 10),
                                        child: DropdownButton(
                                            isExpanded: true,
                                            underline: Container(),
                                            items: [
                                              DropdownMenuItem(
                                                child: Row(
                                                  children: [
                                                    CircleAvatar(
                                                      radius: 15,
                                                      backgroundImage:
                                                          AssetImage('assets/airteltigo.jpg', package: 'harbor_pay'),
                                                    ),
                                                    SizedBox(
                                                      width: 10,
                                                    ),
                                                    appText('AirtelTigo', 15, FontWeight.w600),
                                                  ],
                                                ),
                                                value: 'AIR',
                                              ),
                                              DropdownMenuItem(
                                                child: Row(
                                                  children: [
                                                    CircleAvatar(
                                                      radius: 15,
                                                      backgroundImage:
                                                          AssetImage('assets/mtn.png', package: 'harbor_pay'),
                                                    ),
                                                    SizedBox(
                                                      width: 10,
                                                    ),
                                                    appText('MTN', 15, FontWeight.w600),
                                                  ],
                                                ),
                                                value: 'MTN',
                                              ),
                                              DropdownMenuItem(
                                                child: Row(
                                                  children: [
                                                    CircleAvatar(
                                                      radius: 15,
                                                      backgroundImage:
                                                          AssetImage('assets/vodafone.jpg', package: 'harbor_pay'),
                                                    ),
                                                    SizedBox(
                                                      width: 10,
                                                    ),
                                                    appText('Vodafone', 15, FontWeight.w600),
                                                  ],
                                                ),
                                                value: 'VOD',
                                              )
                                            ],
                                            value: selectedMomo,
                                            onChanged: (val) {
                                              setState(() {
                                                this.transactionType = MOMO_PAY_METHOD;
                                                this.transactionPlatform = val.toString();
                                                selectedMomo = val.toString();
                                              });
                                            }),
                                      ),
                                    )
                                  ],
                                ),
                                Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    ElevatedButton(
                                      onPressed: () {
                                        setState(() {
                                          currentStep = 1;
                                          this.prepareTransaction(context, extra: extra);
                                        });
                                      },
                                      style: bStyle(this.buttonColors, radius: 50),
                                      child: appText('Make Payment', 15, FontWeight.w600, col: Colors.white),
                                    ),
                                  ],
                                ),
                              ],
                            )),
                      ),
                      Step(
                        isActive: currentStep == 1,
                        title: Text('Finish ${this.transactionType}'),
                        content: Container(
                            height: height,
                            child: Column(
                              children: [
                                Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    Padding(
                                      padding: const EdgeInsets.all(10.0),
                                      child: appText('Mobile Money Payment', 15, FontWeight.bold),
                                    )
                                  ],
                                ),
                                Row(
                                  children: [
                                    Expanded(
                                      child: Padding(
                                        padding: const EdgeInsets.all(10.0),
                                        child: appText(
                                            'Confirm the mobile money prompt (if any). Please check your approvals if nothing happens soon.',
                                            15,
                                            FontWeight.normal,
                                            align: TextAlign.center),
                                      ),
                                    )
                                  ],
                                ),
                                SizedBox(height: 20),
                                Center(
                                  child: progress(size: 30),
                                )
                              ],
                            )),
                      ),
                    ],
                    controlsBuilder: (BuildContext context, ControlsDetails details) {
                      return Container();
                    },
                  )));
        });
  }

  headers() {
    var bytes = utf8.encode(this.clientId);
    var auth = base64.encode(bytes);
    return {"Authorization": "Bearer $auth"};
  }
}
